lib/analyse.ml

```diff
@@ -93,26 +100,76 @@ module Analysis = struct
   let ocaml_major_version vars =
     Ocaml_version.with_just_major_and_minor (Ocaml_version.of_string_exn vars.Worker.Vars.ocaml_version)
 
+  let find_compiler ~platforms ~version =
+    let v =
+      Ocaml_version.with_just_major_and_minor
+        (Ocaml_version.of_string_exn version)
+    in
+    let matches_compiler (variant, vars) =
+      if Ocaml_version.equal v (ocaml_major_version vars) then
+        Some variant
+      else
+        None
+    in
+    List.find_map matches_compiler platforms
+
   let duniverse_selections ~job ~platforms dir =
-    let vs = get_ocaml_versions dir in
+    let vs = duniverse_get_ocaml_versions dir in
     if vs = [] then failwith "No OCaml compilers specified!";
-    let find_compiler v =
-      let v = Ocaml_version.with_just_major_and_minor (Ocaml_version.of_string_exn v) in
-      let matches_compiler (_variant, vars) =
-        Ocaml_version.compare v (ocaml_major_version vars) = 0
-      in
-      match List.find_opt matches_compiler platforms with
-      | None ->
-        Current.Job.log job "WARNING: Unsupported compiler version %a (have: %a)"
-          Ocaml_version.pp v Fmt.(Dump.list pp_ocaml_version) platforms;
-        None
-      | Some (variant, _vars) ->
-        Some variant
+    let find_a_compiler version =
+      let r = find_compiler ~version ~platforms in
+      if Option.is_none r then
+        Current.Job.log job "WARNING: Unsupported compiler version %s (have: %a)"
+          version Fmt.(Dump.list pp_ocaml_version) platforms;
+      r
     in
-    let variants = List.filter_map find_compiler vs in
+    let variants = List.filter_map find_a_compiler vs in
     if variants = [] then failwith "No supported compilers found!";
     Lwt_result.return (`Duniverse variants)
 
+  let packages_in_depends f =
+    let get_atom =
+      function
+      | OpamFormula.Atom a -> a
+      | _ -> assert false
+    in
+    OpamFormula.ands_to_list f |> List.map get_atom
+
+  let version_of_equal_constraint f =
+    match f with
+     | OpamFormula.Atom (OpamTypes.Constraint (`Eq, OpamTypes.FString s)) ->
+       s
+     | _ -> assert false
+
+  let opam_monorepo_dep_version ~lock_file ~package =
+    OpamFile.OPAM.depends lock_file
+    |> packages_in_depends
+    |> List.assoc (OpamPackage.Name.of_string package)
+    |> version_of_equal_constraint
+
+  let x_opam_monorepo_version opam =
+    try
+    OpamFile.OPAM.extended opam "x-opam-monorepo-version" (function
+      | String (_, s) -> s
+      | _ -> assert false
+        )
+    with Invalid_argument _ -> None
```

Comment:
Minor: please indent everything with ocp-indent.

Reply:
sure. is it ok to enable ocamlformat on new files? (I see there's an `.ocamlformat-enable` file)

Reply:
I think that enable file is left over from a previous experiment; it should probably be deleted.

I don't want to mandate any particular formatting tool (although I said "ocp-indent" above, I really just mean that the indentation should be consistent; feel free to use ocamlformat on your new file locally before submitting it).

Reply:
ok! it seems that I can't write ocaml without a formatter these days. I'll add a todolist item to format things once the rest is ready.